feat: support HDFS object store

hfutatzhanghb opened the pull request:
## What changed

- add an optional `hdfs` feature for `lance-io` backed by OpenDAL
- register `hdfs://` with the object store registry
- map HDFS URLs, storage options, and supported environment variables into OpenDAL configuration
- add unit coverage for HDFS path and configuration handling

## Why

This allows Lance datasets to be accessed through HDFS URLs, including deployments using HDFS nameservices.

## Validation

- `cargo fmt --all -- --check`
- `cargo clippy -p lance-io --all-targets --features hdfs --locked -- -D warnings`
- `cargo clippy --all --tests --benches -- -D warnings`
- `cargo test -p lance-io --features hdfs --locked object_store::providers::hdfs::tests --no-fail-fast`
- `cargo +1.91.0 check -p lance-io --features hdfs --locked`

Xuanwo commented on rust/lance-io/Cargo.toml:
```diff
@@ -63,13 +64,14 @@ name = "scheduler"
 
 [features]
-default = ["aws", "azure", "gcp"]
+default = ["aws", "azure", "gcp", "hdfs"]
```
I think we shouldn't enable HDFS by default. It introduces many new dependencies in Lance and also requires users to have a Java setup. Without it, Lance will fail to start.

Commit:
feat: lance supports hdfs scheme

hfutatzhanghb commented:
@Xuanwo @BubbleCal @jiaoew1991 Hi， could you please review this PR when free? Thanks very much. We have use HDFS as backend storage for a long time.

Commit:
fix: exclude hdfs from CI all-features builds and update Cargo.lock

The hdfs feature requires Hadoop native libraries (libhdfs.so) which are not
available on CI runners. Exclude it from ALL_FEATURES computation, following
the same pattern as protoc and slow_tests.

Also update Cargo.lock to include hdfs-related dependencies (hdrs, hdfs-sys,
java-locator, opendal-service-hdfs) so --locked builds don't fail when hdfs
is accidentally enabled.

Commit:
fix: improve optional HDFS handling

Commit:
fix java build failed

Commit:
docs: explain Java HDFS builds

Commit:
docs: use mvn in Java build examples

Commit:
docs: document HDFS object store setup

Commit:
docs: move HDFS configuration after TOS

Commit:
optimize feature hdfs

hfutatzhanghb commented:
@yanghua Hi， could you please help review this PR when have free time? Thanks very much!

yanghua commented:
@claude review

hfutatzhanghb commented:
> @claude review
😂， No output ? @claude

yanghua commented:
It seems the bot does not work. I could take a look at this PR later.

hfutatzhanghb commented:
> It seems the bot does not work. I could take a look at this PR later.

Thanks！

yanghua commented on rust/lance-io/src/object_store/providers/hdfs.rs:
```diff
@@ -0,0 +1,268 @@
+            block_size,
+            max_iop_size: *DEFAULT_MAX_IOP_SIZE,
+            use_constant_size_upload_parts: params.use_constant_size_upload_parts,
+            list_is_lexically_ordered: params.list_is_lexically_ordered.unwrap_or(true),
```
Should it be set to `false`?

hfutatzhanghb replied:
I think set to true here is ok.  Because the  output of list rpc in hdfs is lexically_ordered.

yanghua commented on rust/lance-io/src/object_store/providers/hdfs.rs:
```diff
@@ -0,0 +1,268 @@
+                } else {
+                    Some((key.as_ref().to_string(), value))
+                }
+            })
+            .collect::<HashMap<_, _>>();
+
+        let name_node = storage_options
+            .0
+            .get("hdfs_name_node")
+            .cloned()
+            .or_else(|| env_vars.get("HDFS_NAME_NODE").cloned())
+            .unwrap_or_else(|| format!("hdfs://{}", base_path.authority()));
+
+        let mut config = HashMap::from([
+            ("name_node".to_string(), name_node),
+            ("root".to_string(), "/".to_string()),
+        ]);
+
+        let user = storage_options
+            .0
+            .get("hdfs_user")
+            .cloned()
+            .or_else(|| env_vars.get("HADOOP_USER_NAME").cloned())
+            .or_else(|| env_vars.get("HDFS_USER").cloned());
+        if let Some(user) = user {
+            config.insert("user".to_string(), user);
+        }
```
Shall we also filter empty string values?

hfutatzhanghb replied:
Thanks for reminding, Have fixed.

yanghua commented on rust/lance-io/tests/hdfs_integration.rs:
```diff
@@ -0,0 +1,167 @@
+    async fn test_hdfs_store_creation() {
+        if !hdfs_available() {
+            return;
+        }
```
These conditions would make these test cases hard to find real problems.

hfutatzhanghb replied:
Updated the HDFS integration tests to follow the existing object-store integration test pattern used by GooseFS and GCS. Each test is now marked with `#[ignore = "Requires HDFS cluster"]` instead of silently returning based on environment variables. The tests are explicitly run with `--ignored`, and once selected they execute fully so connection or behavior failures are surfaced. The default targeted run now reports `0 passed, 4 ignored` rather than treating unexecuted tests as passing.

yanghua commented on rust/lance-io/tests/hdfs_integration.rs:
```diff
@@ -0,0 +1,167 @@
+    }
+
+    #[tokio::test]
+    async fn test_hdfs_url_parsing() {
```
Is it duplicated with `test_hdfs_store_paths`?

hfutatzhanghb replied:
Yes, have removed test_hdfs_url_parsing. Thanks a lot!

yanghua commented on rust/lance-io/tests/hdfs_integration.rs:
```diff
@@ -0,0 +1,167 @@
+
+        // This test verifies the configuration validation logic without actually connecting
+        // The actual connection would fail, but we're testing the config validation
+        assert!(params.storage_options().is_some());
```
It's too simple, we need to check `HdfsStoreProvider::build_config`.

hfutatzhanghb replied:
Removed the ineffective integration test, which only asserted that `storage_options` was present and did not exercise HDFS configuration parsing. `HdfsStoreProvider::build_config` is private and is tested directly in `rust/lance-io/src/object_store/providers/hdfs.rs`, covering URI fallback, storage option precedence over environment/URI, environment configuration, optional Kerberos/atomic-write options, and rejection of URIs without a host. The targeted HDFS provider unit tests pass (6 tests).

yanghua commented on rust/lance-io/tests/hdfs_integration.rs:
```diff
@@ -0,0 +1,167 @@
+    use url::Url;
+
+    fn hdfs_available() -> bool {
+        std::env::var("HDFS_NAME_NODE").is_ok() || std::env::var("HDFS_TEST_ENABLED").is_ok()
```
Can we add some description on the doc for `HDFS_TEST_ENABLED `?

hfutatzhanghb replied:
Have added. Thanks.

yanghua reviewed (commented):
Left some comments.

Commit:
fix: filter empty string values in HDFS storage options

Add .filter(|v| !v.is_empty()) to hdfs_name_node and hdfs_user
storage options lookups for consistency with the kerberos and
atomic_write_dir fields.

Commit:
refactor: remove duplicate test_hdfs_url_parsing integration test

This test was a duplicate of test_hdfs_store_paths in the unit test
file, both exercise extract_path logic with the same URL patterns.

Commit:
docs: explain HDFS integration test opt-in

Commit:
test: remove ineffective HDFS config validation

Commit:
test: explicitly ignore HDFS integration tests

hfutatzhanghb commented:
@yanghua Thanks very much for valuable review opinions. Have uploaded the latest codes.

yanghua commented:
@hfutatzhanghb, fix the conflicts?

hfutatzhanghb commented:
> @hfutatzhanghb, fix the conflicts?

have fixed and pushed. Let's wait pipeline ok. Thanks for your reviewing!!!

yanghua commented:
Will take a look today.

yanghua reviewed (approved):
+1 cc @Xuanwo

Xuanwo reviewed (changes requested):
Thanks for adding HDFS support. I think this needs a few correctness fixes before merge.

The main blocker is that this PR registers `hdfs://` as writable, but commit routing still falls through to `UnsafeCommitHandler`. That means concurrent HDFS writers can overwrite manifest versions without real create-if-absent / conflict detection semantics.

There are also two HDFS identity/listing issues:
- HDFS is advertised as `list_is_lexically_ordered=true`, but the backend does not guarantee sorted listings before Lance uses listing order for manifest discovery.
- `store_prefix` is derived from the URI authority even when `hdfs_name_node` / `HDFS_NAME_NODE` overrides the actual NameNode, so different effective clusters can be treated as the same datastore.

Please route HDFS to a safe commit handler or explicitly reject HDFS writes without one, default HDFS listings to unordered unless sorted in the provider, and derive `store_prefix` from the resolved effective NameNode.

hfutatzhanghb commented:
> Thanks for adding HDFS support. I think this needs a few correctness fixes before merge.
> 
> The main blocker is that this PR registers `hdfs://` as writable, but commit routing still falls through to `UnsafeCommitHandler`. That means concurrent HDFS writers can overwrite manifest versions without real create-if-absent / conflict detection semantics.
> 
> There are also two HDFS identity/listing issues:
> 
> * HDFS is advertised as `list_is_lexically_ordered=true`, but the backend does not guarantee sorted listings before Lance uses listing order for manifest discovery.
> * `store_prefix` is derived from the URI authority even when `hdfs_name_node` / `HDFS_NAME_NODE` overrides the actual NameNode, so different effective clusters can be treated as the same datastore.
> 
> Please route HDFS to a safe commit handler or explicitly reject HDFS writes without one, default HDFS listings to unordered unless sorted in the provider, and derive `store_prefix` from the resolved effective NameNode.

@Xuanwo Thanks for reviewing.
1. For UnsafeCommitHandler,  we have already implements a HdfsRenameCommitHandler and planed to push in another PR after this mergeed.  Now have pushed in this PR. 
2. For indentity/listing issues: Will fix immediately.

Commit:
fix: set list_is_lexically_ordered to false for HDFS

Commit:
fix: derive store_prefix from resolved effective NameNode for HDFS

Commit:
feat: add safe HDFS rename commit handler via libhdfs

Commit:
style: fix cargo fmt issues in hdfs commit handler

Commit:
test: add unit tests for HDFS commit handler helper functions

hfutatzhanghb commented:
@Xuanwo  @yanghua  To provide better rename semantics for the Lance HDFS feature, we also optimized the libhdfs rename operation in https://github.com/apache/hadoop/pull/8411.

Commit:
fix: move StorageOptionsAccessor import into test module

Commit:
chore: sync python/Cargo.lock with workspace hdrs dep

Xuanwo commented on Cargo.toml:
```diff
@@ -104,6 +104,7 @@ half = { "version" = "2.1", default-features = false, features = [
     "num-traits",
     "std",
 ] }
+hdrs = "0.3.2"
```
We should not depend on this

hfutatzhanghb replied:
Hi, @Xuanwo . Do you mean we should avoid using hdrs directly and implement the HDFS commit rename through OpenDAL instead?  My concern is preserving the rename-if-destination-does-not-exist semantics required for commit conflict detection.

hfutatzhanghb replied:
OpenDAL’s HDFS rename currently overwrites the destination by deleting it first. The manifest commit requires an atomic rename-if-destination-does-not-exist operation for conflict detection. Would you prefer adding conditional rename support to OpenDAL first, or is there another OpenDAL API intended for this semantic?

wjones127 replied:
See me comment here: https://github.com/lance-format/lance/issues/7584#issuecomment-4898152245

hfutatzhanghb replied:
Hi, @wjones127 . Got it. That a good plan. This PR is not the final version currently. It depends on the rename_with capabilty of OpenDAL which was introduced by https://github.com/apache/opendal/pull/7815 . We should release a new version for OpenDAL firstly.   BTW, in our cluster. We use patch.crates-io to patch the OpenDAL code and HDRS code.

Commit:
fix: use opendal for hdfs commits
